include/nbl/video/utilities/IGPUObjectFromAssetConverter.h

```diff
@@ -1752,22 +1766,14 @@ inline created_gpu_object_array<asset::ICPUDescriptorSet> IGPUObjectFromAssetCon
                         {
                             info->desc = imgViewRedirs[ivi] >= gpuImgViews->size() ? nullptr : gpuImgViews->operator[](imgViewRedirs[ivi]);
                             ++ivi;
-                            // TODO: This should be set in the loader (or whoever is creating
-                            // the descriptor)
-                            if (info->info.image.imageLayout == asset::IImage::EL_UNDEFINED)
+                            info->info.image.imageLayout = descriptorInfos[offset + d].info.image.imageLayout;
+                            assert(info->info.image.imageLayout != asset::IImage::EL_UNDEFINED);
+
+                            if (!isStorageImgDesc(type))
                             {
-                                if (isStorageImgDesc(type))
-                                {
-                                    info->info.image.imageLayout = asset::IImage::EL_GENERAL;
-                                }
-                                else
-                                {
-                                    const auto imageFormat = static_cast<asset::ICPUImageView*>(info->desc.get())->getCreationParameters().format;
-                                    info->info.image.imageLayout = isDepthOrStencilFormat(imageFormat) ? asset::IImage::EL_DEPTH_STENCIL_READ_ONLY_OPTIMAL : asset::IImage::EL_SHADER_READ_ONLY_OPTIMAL;
-
-                                    if (samplers.size() != 0ull)
-                                        info->info.image.sampler = gpuSamplers->operator[](smplrRedirs[si++]);
-                                }
+                                const bool isMutableSamplerBinding = (mutableSamplerBindingRedirect.searchForBinding(asset::ICPUDescriptorSetLayout::CBindingRedirect::binding_number_t{ write_it->binding }) != mutableSamplerBindingRedirect.Invalid);
+                                if (descriptorInfos.begin()[offset + d].info.image.sampler && isMutableSamplerBinding)
+                                    info->info.image.sampler = gpuSamplers->operator[](smplrRedirs[si++]);
```

Comment:
you probably just want to predicate/check base on the layout

Reply:
Are you saying that I should remove the `descriptorInfos.begin()[offset + d].info.image.sampler` condition from the `if`?

Reply:
Yes, `assert`-ing is better. Avoids silent fails.
Resolved.